kvdb-rocksdb/benches/bench_read_perf.rs

```diff
@@ -0,0 +1,213 @@
+/// an `ElasticArray128` so sometimes we save on allocations.
+fn n_random_bytes(n: usize) -> Vec<u8> {
+	let mut rng = rand::thread_rng();
+	let variability: i64 = rng.gen_range(0, (n as f64 * 0.2) as i64);
+	let plus_or_minus: i64 = if variability % 2 == 0 { 1 } else { -1 };
+	let range = Uniform::from(0..u8::max_value());
+	rng.sample_iter(&range).take((n as i64 + plus_or_minus * variability) as usize).collect()
+}
+
+/// Writes `NEEDLES * NEEDLES_TO_HAYSTACK_RATIO` keys to the DB. Keys are random, 32 bytes long and
+/// values are random, 120-180 bytes long. Every `NEEDLES_TO_HAYSTACK_RATIO` keys are kept and
+/// returned in a `Vec` for and used to benchmark point lookup performance. Keys are sorted
+/// lexicographically in the DB, and the benchmark keys are random bytes making the needles are
+/// effectively random points in the key set.
+fn populate(db: &Database) -> io::Result<Vec<H256>> {
+	let mut needles = Vec::with_capacity(NEEDLES);
+	let mut batch = db.transaction();
+	for i in 0..NEEDLES * NEEDLES_TO_HAYSTACK_RATIO {
+		let key = H256::random();
+		if i % NEEDLES_TO_HAYSTACK_RATIO == 0 {
+			needles.push(key.clone());
+			if i % 100_000 == 0 && i > 0 {
+				println!("[populate] {} keys", i);
+			}
+		}
+		// In ethereum keys are mostly 32 bytes and payloads ~140bytes.
+		batch.put(Some(0), &key.as_bytes(), &n_random_bytes(140));
+	}
+	db.write(batch)?;
+	// Clear the overlay
+	db.flush()?;
+	Ok(needles)
+}
+
+fn get(c: &mut Criterion) {
+	let db = open_db();
+	let needles = populate(&db).expect("rocksdb works");
+
+	let mut total_iterations = 0;
+	let mut total_allocs = 0;
+
+	c.bench_function("get key", |b| {
+		b.iter_custom(|iterations| {
+			total_iterations += iterations;
+			let mut elapsed = Duration::new(0, 0);
+			// NOTE: counts allocations on the Rust side only
+			let (alloc_stats, _) = count_alloc(|| {
+				let start = Instant::now();
+				for _ in 0..iterations {
+					// This has no measurable impact on performance (~30ns)
+					let needle = needles.choose(&mut rand::thread_rng()).expect("needles is not empty");
+					black_box(db.get(Some(0), needle.as_bytes()).unwrap());
+				}
+				elapsed = start.elapsed();
+			});
+			total_allocs += alloc_stats.0;
+			elapsed
+		});
+	});
+	if total_iterations > 0 {
+		println!(
+			"[get key] total: iterations={}, allocations={}; allocations per iter={:.2}\n",
+			total_iterations,
+			total_allocs,
+			total_allocs as f64 / total_iterations as f64
+		);
+	}
+
+	total_iterations = 0;
+	total_allocs = 0;
+	c.bench_function("get key by prefix", |b| {
+		b.iter_custom(|iterations| {
+			total_iterations += iterations;
+			let mut elapsed = Duration::new(0, 0);
+			// NOTE: counts allocations on the Rust side only
+			let (alloc_stats, _) = count_alloc(|| {
+				let start = Instant::now();
+				for _ in 0..iterations {
+					// This has no measurable impact on performance (~30ns)
+					let needle = needles.choose(&mut rand::thread_rng()).expect("needles is not empty");
```

Comment:
move rand::thread_rng() outside the loop and closure?

Reply:
It's a lazily-initialized global (per thread) object, https://docs.rs/rand/0.7.2/rand/fn.thread_rng.html

Reply:
(so no need to move it out of the loop)